R/bs-theme.R

```diff
@@ -141,25 +152,32 @@ bs_theme <- function(version = version_default(), bootswatch = NULL, ...,
 #' @rdname bs_theme
 #' @param theme a [bs_theme()] object.
 #' @export
-bs_theme_update <- function(theme, ..., bootswatch = NULL, bg = NULL, fg = NULL,
+bs_theme_update <- function(theme, ..., name = NULL, bg = NULL, fg = NULL,
                             primary = NULL, secondary = NULL, success = NULL,
                             info = NULL, warning = NULL, danger = NULL,
                             base_font = NULL, code_font = NULL, heading_font = NULL,
-                            font_scale = NULL) {
+                            font_scale = NULL, bootswatch = NULL) {
   assert_bs_theme(theme)
 
-  if (!is.null(bootswatch)) {
-    old_swatch <- theme_bootswatch(theme)
-    # You're only allowed one Bootswatch theme!
-    if (length(old_swatch)) {
-      theme <- bs_remove(theme, "bootswatch")
-      class(theme) <- setdiff(class(theme), bootswatch_class(old_swatch))
-    }
-    if (!identical(bootswatch, "default")) {
-      theme <- add_class(theme, bootswatch_class(bootswatch))
-      theme <- bs_bundle(theme, bootswatch_bundle(bootswatch, theme_version(theme)))
+  theme_has_preset <- any(grepl("^bs_(builtin|bootswatch)_", class(theme)))
+
+  preset <- resolve_bs_preset(name, bootswatch, version = theme_version(theme))
+
+  if (!is.null(preset)) {
+    if (theme_has_preset) {
+      old_preset_class <- grep("^bs_(builtin|bootswatch)_", class(theme), value = TRUE)
+      old_preset_type <- sub("^bs_(builtin|bootswatch)_.+", "\\1", old_preset_class)
```

Comment:
Not today, but the S3 methods in bs-theme-preset.R really invite extending the theme object. 

Would it be possible that code paths like these that are detecting a set of objects be made into S3 methods? 

Motivation: It could allow for pharma to make their own bundles to be swappable. 

---------------------



Or could the class name be stored within the object (or be able to be created from the existing `preset` object? Then we are not matching the class name here

Could this logic be put into a method like `has_preset_theme()`? And also add a `theme_remove_preset()` (or `bs_preset_reset()`)? (IDK the proper namespacing for these methods. :-/)

Reply:
I'll talk about this with @cpsievert when he's back in the office. I like the idea of making a theme reset generic that uses S3 classes to remove the preset bundle. I'm also suspicious of storing the preset name in the class, but I'm assuming Carson had reasons for this approach (and guessing that its (a) related to the structure of the sass bundle object and (b) may be safer to use class names than to attach attributes?)